docker/Dockerfile

```diff
@@ -88,24 +86,136 @@ RUN apt-get update && \
 	# Remove Git and clean apt cache
 	apt-get clean && \
 	apt-get remove -y git && \
-	apt-get -y autoremove && \
-	chmod +x docker/entrypoint.sh
+	apt-get -y autoremove
+
+
+
+########################################################################################
+#
+# This build stage installs PM2 if required.
+# 
+# To create an image using PM2 run this Docker file with the parameter:
+#
+# --target=pm2-install
+#
+########################################################################################
+FROM node:${NODE_VERSION}-slim AS pm2
+ARG GITHUB_ACCOUNT
+ARG GITHUB_REPOSITORY
+ARG NODE_VERSION
+
+LABEL "maintainer"="FIWARE IoTAgent Team. Telefónica I+D"
+LABEL "org.opencontainers.image.authors"="iot_support@tid.es"
+LABEL "org.opencontainers.image.documentation"="https://fiware-iotagent-lwm2m.rtfd.io/"
+LABEL "org.opencontainers.image.vendor"="Telefónica Investigación y Desarrollo, S.A.U"
+LABEL "org.opencontainers.image.licenses"="AGPL-3.0-only"
+LABEL "org.opencontainers.image.title"="FIWARE IoT Agent for OMA LightWeight Machine2Machine with PM2"
+LABEL "org.opencontainers.image.description"="IoT Agent accepting COAP requests. Designed to be a bridge between the OMA Lightweight M2M protocol and the NGSI interface."
+LABEL "org.opencontainers.image.source"="https://github.com/${GITHUB_ACCOUNT}/${GITHUB_REPOSITORY}"
+LABEL "org.nodejs.version"="${NODE_VERSION}"
+
+COPY --from=builder /opt/iota-lwm2m /opt/iota-lwm2m
+RUN npm install pm2@4.4.0 -g --no-package-lock --no-optional
 
 USER node
-ENV NODE_ENV=production
+ENV NODE_ENV=development
+# Expose 4041 for NORTH PORT, 5683 for LWM2M PORT
+EXPOSE ${IOTA_NORTH_PORT:-4041} ${LWM2M_PORT:-5683}/udp
+CMD ["pm2-runtime", "/opt/iota-lwm2m/bin/lwm2mAgent.js", "-- ", "config.js"]
 
-# Expose Ports - Default to 4041 for NORTH PORT, 5683 for LWM2M PORT
+########################################################################################
+#
+# This build stage creates an anonymous user to be used with the distroless build
+# as defined below.
+#
+########################################################################################
+FROM node:${NODE_VERSION} AS anon-user
+RUN sed -i -r "/^(root|nobody)/!d" /etc/passwd /etc/shadow /etc/group \
+    && sed -i -r 's#^(.*):[^:]*$#\1:/sbin/nologin#' /etc/passwd
+
+########################################################################################
+#
+# This build stage creates a distroless image for production.
+#
+# IMPORTANT: For production environments use Docker Secrets to protect values of the 
+# sensitive ENV variables defined below, by adding _FILE to the name of the relevant 
+# variable.
+#
+# - IOTA_AUTH_USER, IOTA_AUTH_PASSWORD - when using Keystone Security 
+# - IOTA_AUTH_CLIENT_ID, IOTA_AUTH_CLIENT_SECRET - when using OAuth2 Security
+#
+########################################################################################
+FROM gcr.io/distroless/nodejs:${NODE_VERSION}
```

Comment:
Looking to "reference" Dockerfile at https://github.com/telefonicaid/iotagent-json/blob/master/docker/Dockerfile#L155 maybe a "AS distroless" is missing here?

Reply:
Fixed 66e2b6b